File: scripts/framework/net/KcpServer.lua
net = net or {}

function net.NewKCPBinaryServer(address, tag)
    local self = {
    }
    local clients = {}
    local ptr = nil
    local function onNew(conn)
        local cli = {
            conn = conn,
        }
        function cli.Send(data)
            BabyEngine.Net.Send(ptr, cli.conn, data)
        end
        function cli.Close()
            BabyEngine.Net.Close(ptr, cli)
        end
        function cli.Redirect(address)
            BabyEngine.Net.Redirect(ptr, cli, address)
        end
        clients[conn] = cli
        if not self.OnNew then return end
        self.OnNew(cli)
    end
    local function onClose(conn)
        local cli = clients[conn]
        if not cli then return end
        clients[conn] = nil
        if not self.OnClose then return end
        self.OnClose(cli)
    end
    local function onError(conn, err)
        local cli = clients[conn]
        if not cli then return end
        if not self.OnError then return end
        self.OnError(cli, err)
    end
    local function onData(conn, data)
        local cli = clients[conn]
        if not cli then return end
        if not self.OnData then return end
        self.OnData(cli, data)
    end
    local n = 0
    local function onRequest(conn, data)
        local cli = clients[conn]
        if not cli then return end
        if not self.OnRequest then return end
        return self.OnRequest(cli, data)
    end
    function self.Start( )
        ptr = BabyEngine.Net.Start('kcp', address, tag)
        BabyEngine.Net.Bind(ptr, "new",  onNew)
        BabyEngine.Net.Bind(ptr, "close",  onClose)
        BabyEngine.Net.Bind(ptr, "error",  onError)
        BabyEngine.Net.Bind(ptr, "data",  onData)
        BabyEngine.Net.Bind(ptr, "request",  onRequest)
    end

    function self.Stop()
        if ptr then
            BabyEngine.Net.Stop(ptr)
            ptr = nil
        end
    end
    return self
end

File: networking/kcp_server.go
package networking

import (
    "fmt"
    "github.com/BabyEngine/Backend/logger"
    "github.com/xtaci/kcp-go"
    "net"
    "sync"
    "sync/atomic"
    "time"
)

type mKCPServer struct {
    opts     *Options
    wg       sync.WaitGroup
    totalQPS uint32
    tx       uint64 // transfer bytes
    rx       uint64 // receive bytes
    tp       uint64 // transfer packet
    rp       uint64 // receive packet
}

func (s *mKCPServer) Init() {
}

func (s *mKCPServer) Serve(address string) error {
    var (
        retErr error
    )
    ln, err := kcp.Listen(address)
    if err != nil {
        return err
    }
    ticker := time.NewTicker(time.Second)
    defer func() {
        if ln != nil {
            if err := ln.Close(); err != nil {
                logger.Debug(err)
            }
        }
        ticker.Stop()
    }()
    quitChan := make(chan bool)
    go func() {
    OUT:
        for {
            select {
            case <-s.opts.Ctx.Done():
                if ln != nil {
                    if err := ln.Close(); err != nil {
                        logger.Debug(err)
                    }
                }
                ln = nil
            case <-ticker.C:
                s.checkClient()
            case <-quitChan:
                break OUT
            }
        }
    }()

    for {
        conn, err := ln.Accept()
        if err != nil {
            if ne, ok := err.(net.Error); ok && ne.Temporary() {
                // ignore code here
                continue
            }
            logger.Error(err)
            retErr = err
            break
        }
        go s.handleKCPConn(conn)
    }
    quitChan <- true
    return retErr
}

func (s *mKCPServer) handleKCPConn(conn net.Conn) {
    client := &mKCPClient{
        conn:   conn,
        opts:   s.opts,
        server: s,
    }

    client.init()
    logger.Debugf("accept kcp client: %s", client)

    s.wg.Add(1)
    defer func() {
        s.wg.Done()
        s.opts.Handler.OnClose(client)
    }()

    client.Serve()
}

func (s *mKCPServer) checkClient() {
    var (
        checkList []*mKCPClient
        deathList []*mKCPClient
    )
    if s.opts == nil || s.opts.Handler == nil {
        return
    }
    clients := s.opts.Handler.GetAllClient()
    for _, cli := range clients {
        if cc, ok := cli.(*mKCPClient); ok {
            checkList = append(checkList, cc)
        }
    }

    for _, cli := range checkList {
        if !cli.IsAlive() {
            deathList = append(deathList, cli)
        }
    }

    if len(deathList) > 0 {
        for _, cli := range deathList {
            s.opts.Handler.OnClose(cli)
        }
    }
    totalCount := len(clients)
    deathCount := len(deathList)
    if deathCount > 0 {
        fmt.Printf("当前客户端 总数:%d 死亡:%d\n", totalCount, deathCount)
    }
}

// 网络统计
func (s *mKCPServer) onNetStat(aType int, n, l uint64) {
    if aType == 0 {
        atomic.AddUint64(&s.rp, n)
        atomic.AddUint64(&s.rx, l)
    } else if aType == 1 {
        atomic.AddUint64(&s.tp, n)
        atomic.AddUint64(&s.tx, l)
    }
}
